src/aof.c

```diff
@@ -931,6 +1405,108 @@ int loadAppendOnlyFile(char *filename) {
+     * there is only one possibility: the aof manifest file does not exist, we may 
+     * be starting from an old redis version. so we must fall back to the previous 
+     * loading mode.
+     */
+    if (am->base_aof_info == NULL && listLength(am->incr_aof_list) == 0) {
+        struct redis_stat sb;
+        /* If the `server.aof_filename` file does not exist, we will directly return 
+         * AOF_NOT_EXIST, and redis will ignore this error. */
+        if (redis_stat(server.aof_filename, &sb) != 0) {
+            return AOF_NOT_EXIST;
+        }
+
+        /* If the `server.aof_filename` file exists, we manually construct a BASE 
+         * type aofInfo and add it to aofManifest. In this way, we can reuse the 
+         * following code to load this file. */
+        aofInfo *ai = aofInfoCreate();
+        ai->file_name = sdsnew(server.aof_filename);
+        ai->file_seq = 1;
+        ai->file_type = AOF_FILE_TYPE_BASE;
+        am->base_aof_info = ai;
+        am->curr_base_file_seq = 1;
+        am->dirty = 1;
+
+        int ret = persistAofManifest(am);
+        if (ret == C_ERR) {
+            return AOF_FAILED;
+        }
+    }
+
+    serverAssert(am->base_aof_info != NULL || listLength(am->incr_aof_list) != 0);
+
+    /* Here we calculate the total size of all BASE and INCR files in 
+     * advance, it will be set to `server.loading_total_bytes`. */
+    total_size = getBaseAndIncrAppendOnlyFilesSize(server.aof_manifest);
+    startLoading(total_size, RDBFLAGS_AOF_PREAMBLE, 0);
+
+    /* Load BASE AOF if needed. */
+    if (am->base_aof_info) {
+        serverAssert(am->base_aof_info->file_type == AOF_FILE_TYPE_BASE);
+        aof_name = (char*)am->base_aof_info->file_name;
+        updateLoadingFileName(aof_name);
+
+        start = ustime();
+        ret = loadSingleAppendOnlyFile(aof_name);
+        if (ret == AOF_OK) {
+            serverLog(LL_NOTICE, "DB loaded from base file %s: %.3f seconds",
+                aof_name, (float)(ustime()-start)/1000000);
+        }
+
+        /* If an AOF exists in the manifest but not on the disk, we consider 
+         * this to be a fatal error. */
+        if (ret == AOF_NOT_EXIST) ret = AOF_FAILED;
+
+        if (ret != AOF_OK && ret != AOF_EMPTY) {
+            goto cleanup;
+        }
+    }
+
+    /* Load INCR AOFs if needed. */
+    if (listLength(am->incr_aof_list)) {
+        listNode *ln;
+        listIter li;
+
+        listRewind(am->incr_aof_list, &li);
+        while ((ln = listNext(&li)) != NULL) {
+            aofInfo *ai = (aofInfo*)ln->value;
+            serverAssert(ai->file_type == AOF_FILE_TYPE_INCR);
+            aof_name = (char*)ai->file_name;
+            updateLoadingFileName(aof_name);
+
+            start = ustime();
+            ret = loadSingleAppendOnlyFile(aof_name);
+            if (ret == AOF_OK) {
+                serverLog(LL_NOTICE, "DB loaded from incr file %s: %.3f seconds", 
```

Comment:
Do we really need this? Currently, we can get the progress of loading through `server.loading_loaded_bytes` and `server.loading_loaded_perc` (print in redis info). In addition, if there is an error in the intermediate loading, we will also print out the file name with the error. I think the information we printed is enough now, @oranagra @yossigo WDYT?

Reply:
the code we're looking at, prints a message when the file is present and was loaded successfully,
and obviously when there's some parsing error, it prints a log message to (inside the function).
besides that i see the function also has a specific print when the file is missing, and when it's empty.
so i think that log-wise we're covered.

i do see that we fail the loading when the file is present and empty, and maybe that's not good?
the old code used to succeed when a file was either missing or empty.
for the new code, we don't expect it to be missing (if we added it to the manifest, it should be there), but it could be empty (if there's no traffic after rewrite started).

maybe the base file must never be empty, since that's either an RDB, or an AOF with at least a SELECT statement?
but what if the server just started (a new deployment with no pre-existing data or persistence files), and then it is restarted after it created the AOF file and didn't yet process any write command?

i suppose the same could also be with upgrades (no manifest file).
and i guess this concern will be resolved once we implement #9794 and always generate an RDB as base (even when starting empty).

Reply:
@oranagra I think you may have misunderstood it. The current implementation is: Only when AOF does not exist will load error, when it is empty, we will directly load the next one, The following is my handling of the return value:
```
         /* If an AOF exists in the manifest but not on the disk, we consider 
         * this to be a fatal error. */
        if (ret == AOF_NOT_EXIST) ret = AOF_FAILED;

        if (ret != AOF_OK && ret != AOF_EMPTY) {
            goto cleanup;
        }
```

Reply:
yes, you're right, i misinterpreted that code.
so we skip empty files, and fail on non existing ones, and either of the 4 states (missing, empty, error, success) there is a log print.
i think we're good, but maybe i'm missing Yossi's intention in the original comment.